### PR TITLE
Adds an option to fill the forecast segment with icons

### PR DESCRIPTION
**Pull request by scinos**
(Comes from #669)

## What problem does this PR solve?

Adds the option to add more than one icon per forecast span. This should improve readability, specially for wallboards.

## Proposed solution

Adds a new config `icon_fill` with three possible values:

- `single` This is the default option and keeps current functionallity: it shows one single icon per span, centered.
- `full` Show one icon per forecast segment. For example, it will display 12 sun icons if the next 12h is sunny.
- `<n>` (an integer). Like `full`, but instead of displaying an icon per forecast segment, it will displaye an icon every nth segment. (In other words, `full` and `1` do the same). At the very least it will display one icon per span.

This is how it looks like:

image

Alternatively, we can have this as a boolean (`icon_fill:false` current behaviour, `icon_fill:true` fill all segments with icons), and introduce a second option `icon_spacing` (similar to `label_spacing`) to be able to space out the icons

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -61,30 +61,31 @@ decimal by 1). Otherwise, the integration may complain of a duplicate unique ID.
...
+| `hide_temperatures`              | bool                | **Optional** | Whether to hide temperatures under the bar                                                                                                                      | `false`             |
+| `round_temperatures`             | bool                | **Optional** | Whether to round temperatures to the nearest whole number                                                                                                       | `false`             |
+| `hide_bar`                       | bool                | **Optional** | Whether to hide the bar itself                                                                                                                                  | `false`             |
+| `icon_fill`                      | [string][icon_fill] | **Optional** | Whether to repeat the icon inside the bar                                                                                                                       | `'single`           |
```
Comment (scinos) [inline README.md:80]: This the only real change in this file. Everything else is adding spaces to keep columns aligned.

File: README.md [hunk 2]
```diff
@@ -205,6 +206,14 @@ colors:
 - `boundary` Show date at the boundary between days
 - `all` Always show date
 
+### Icon Fill Options
+
+`icon_fill` can be one of the following values:
+
+- `single` Show one icon per forecast span (deafult)
```
Comment (decompil3d) [inline README.md:213]:
```suggestion
- `single` Show one icon per forecast span (default)
```

File: README.md [hunk 3]
```diff
@@ -205,6 +206,14 @@ colors:
...
+- `full` Show one icon per forecast segment.
+- `<n>` (an integer). Show icons every _n-th_ forecast segments (will at least show one icon per forecast span)
```
Comment (decompil3d) [inline README.md:215]:
```suggestion
- `<n>` (an integer). Show icons every _n-th_ forecast segment (will show at least one icon per forecast span)
```

File: src/hourly-weather.ts [hunk 4]
```diff
@@ -196,6 +196,11 @@ export class HourlyWeatherCard extends LitElement {
       }
     }
 
+    if (config.icon_fill && config.icon_fill !== 'single' && config.icon_fill !== 'full') {
+      if (!Number.isNaN(config.icon_fill) && (config.icon_fill < 1)) {
+        throw new Error(this.localize('errors.must_be_positive_int'));
```
Comment (decompil3d) [inline src/hourly-weather.ts:201]: Let's not reuse this error string, as it won't be very helpful to users if they mistype the value. Can you make a new string in `src/localize/languages/en.json` under the `errors` section with something like this:

```
icon_fill must be either a positive integer or one of "single" or "full"
```

No need to translate to other languages. Once I merge this, I can trigger auto-translation on the new string before releasing.

File: src/weather-bar.ts [hunk 5]
```diff
@@ -68,12 +71,30 @@ export class WeatherBar extends LitElement {
...
         if (icon === cond[0]) icon = 'mdi:weather-' + icon;
         else icon = 'mdi:' + icon;
+
+        const iconMarkup: TemplateResult[] = [];
+        if (!this.icons) {
+          iconMarkup.push(html`<span class="condition-label">${label}</span>`);
+        } else {
+          let iconSize: IconFillType;
+          if (this.icon_fill === 'single') {
+            iconSize = cond[1]
```
Comment (decompil3d) [inline src/weather-bar.ts:81]:
```suggestion
            iconSize = cond[1]; // grid width of segment, so one icon
```

**Review by decompil3d — COMMENTED** [review]
Thanks for splitting this into a separate PR. Much easier to review. This overall looks good. Some minor comments below. Please address the comments and then add tests for the new behavior.

File: src/hourly-weather.ts (hunk 4 above)
Comment (scinos) [reply src/hourly-weather.ts:201]: Got it. I also refactored the check to be more explicit about what is a valid value and what isn't.

**Comment by scinos** [comment]
@decompil3d I think I need some help with checking the config option `icon_fill`. I see there are config tests in two places:
 - Class `HourlyWeatherCard`, method `setConfig` (https://github.com/decompil3d/lovelace-hourly-weather/blob/main/src/hourly-weather.ts#L192)
 - Class `HourlyWeatherCard`, method `renderCore` (https://github.com/decompil3d/lovelace-hourly-weather/blob/main/src/hourly-weather.ts#L324)

Why do we need two? Which one should I use to test the validity of `icon_fill` option?

**Comment by decompil3d** [comment]
> @decompil3d I think I need some help with checking the config option `icon_fill`. I see there are config tests in two places:
> 
>  - Class `HourlyWeatherCard`, method `setConfig` (https://github.com/decompil3d/lovelace-hourly-weather/blob/main/src/hourly-weather.ts#L192)
> 
>  - Class `HourlyWeatherCard`, method `renderCore` (https://github.com/decompil3d/lovelace-hourly-weather/blob/main/src/hourly-weather.ts#L324)
> 
> 
> 
> Why do we need two? Which one should I use to test the validity of `icon_fill` option?

I think checking validity in `renderCore` would be most consistent with the existing code. Thanks for checking.

**Comment by scinos** [comment]
@decompil3d added tests, this PR is ready for review now

**Review by decompil3d — APPROVED** [review]
Looks good. I'm on my phone reviewing this right now, but I'll merge this into a branch later so that CI picks up translation and then will merge to main and publish. Thanks for your contribution!

**Pull request merged**